### PR TITLE
[FLINK-3777] Managed closeInputFormat() 

### PR DESCRIPTION
**Pull request by fpompermaier**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [X] General
  - The pull request references the related JIRA issue
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message

- [X] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

-  [X] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

--------------

Added Javadocs, don't know wheter other documentation needs to be updated

### CODE REVIEW

**Comment by zentol** [comment]
there are a few formatting issues with missing spaces after if/braces.

**Comment by fhueske** [comment]
Thanks for this PR @fpompermaier.

We need to check all places where `InputFormat.configure()` is called to ensure that we free resources as promised in the JavaDocs of `RichInputFormat.closeInputFormat()`. I found a few more spots where we need to call `closeInputFormat()` if the IF is a `RichInputFormat`.

- ReplicatingInputFormat which is a wrapper for other input formats.
- DataSourceNode.computeOperatorSpecificDefaultEstimates() which tries to obtain some input statistics and initializes the input format with `configure()`.
- `InputFormatVertex.initializeOnMaster()` calls `configure()` as well.

It would also be good to add a test that the closeInputFormat() method is actually called.

File: flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/DataSourceNode.java [hunk 1]
```diff
@@ -150,6 +151,12 @@ protected void computeOperatorSpecificDefaultEstimates(DataStatistics statistics
 				}
 				return;
 			}
+			finally {
+				// close the instantiated resources before returning
+				if(format != null && format instanceof RichInputFormat) {
```
Comment (fhueske) [inline flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/DataSourceNode.java]: If we close the IF here, we might not be able to retrieve the stats.

File: flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/DataSourceNode.java [hunk 2]
```diff
@@ -187,6 +194,10 @@ else if (len >= 0) {
 					this.estimatedNumRecords = card;
 				}
 			}
+			// close the instantiated resources
+			if(format != null && format instanceof RichInputFormat) {
```
Comment (fhueske) [inline flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/DataSourceNode.java]: IF should be closed after the stats have been retrieved (`format.getStatistics(cachedStatistics);`).

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputFormatVertex.java [hunk 3]
```diff
@@ -64,13 +65,19 @@ public void initializeOnMaster(ClassLoader loader) throws Exception {
...
+		InputFormat<?, ?> inputFormat = null;
 		try {
 			inputFormat = wrapper.getUserCodeObject(InputFormat.class, loader);
 		}
 		catch (Throwable t) {
 			throw new Exception("Instantiating the InputFormat (" + formatDescription + ") failed: " + t.getMessage(), t);
 		}
+		finally{
+			// close the instantiated resources before returning
+			if(inputFormat != null && inputFormat instanceof RichInputFormat) {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputFormatVertex.java]: `inputFormat.configure()` has not been called yet. Why are you closing it?

**Comment by fhueske** [comment]
Hi @fpompermaier, I thought about this again. I don't think it is a good idea to tie `closeInputFormat()` to `configure()` because `configure()` is called at different places such as the JobManager and in the optimizer to retrieve statistics. Hence, it is not a good idea to open resources in `configure()`.

We should rather add a `openInputFormat()` method to `RichInputFormat()` which is called directly before `open()`. Would you add such a method and remove the unnecessary `closeInputFormat()` calls, and tests that check that `openInputFormat()` and `closeInputFormat()` are correctly called in the life cycle of a RichInputFormat? 

Sorry for not thinking about this earlier. Thanks, Fabian

**Comment by fpompermaier** [comment]
Me and Stefano agree on this :)
I'll will edit the PR and commit today

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputFormatVertex.java [hunk 4]
```diff
@@ -82,9 +82,6 @@ public void initializeOnMaster(ClassLoader loader) throws Exception {
 		catch (Throwable t) {
 			throw new Exception("Configuring the InputFormat (" + formatDescription + ") failed: " + t.getMessage(), t);
 		}
-		finally {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputFormatVertex.java]: Is this change intended?

File: flink-core/src/main/java/org/apache/flink/api/common/io/ReplicatingInputFormat.java [hunk 5]
```diff
@@ -133,4 +133,18 @@ public RuntimeContext getRuntimeContext(){
 			throw new RuntimeException("The underlying input format to this ReplicatingInputFormat isn't context aware");
 		}
 	}
+	
+	@Override
+	public void openInputFormat() {
+		if(this.replicatedIF instanceof RichInputFormat){
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/io/ReplicatingInputFormat.java]: Can you add a space before and after the parentheses of the condition, i.e., `if (condition) {`?
Please check the other changes as well.

**Comment by fhueske** [comment]
Thanks for the quick update @fpompermaier! The changes look mostly good. 

What is missing are tests that verify that `DataSourceTask` and `GenericDataSourceBase` correctly call the new methods as part of their IF life cycle. Can you extend `GenericDataSourceBaseTest` and `DataSourceTask` and add tests to validate that the methods are called (maybe by using a mocked `RichInputFormat`)? Thanks

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputFormatVertex.java (hunk 4 above)
Comment (fpompermaier) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputFormatVertex.java]: no..sorry :)

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/DataSourceTaskTest.java [hunk 6]
```diff
@@ -84,16 +85,25 @@ public void testDataSourceTask() {
...
 		} catch (Exception e) {
 			System.err.println(e);
 			Assert.fail("Invoke method caused exception.");
 		}
 		
+		try {
+			Field formatField = DataSourceTask.class.getDeclaredField("format");
+			formatField.setAccessible(true);
+			MockInputFormat inputFormat = (MockInputFormat) formatField.get(testTask);
+			Assert.assertTrue("Invalid status of the input format. Expected for closed: true, Actual: "+inputFormat.opened, inputFormat.closed);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/DataSourceTaskTest.java]: missing spaces around +

Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/DataSourceTaskTest.java]: you're also printing ```if.opened``` as the actual value but check ```if.closed```.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java [hunk 7]
```diff
@@ -200,6 +209,7 @@ else if (!this.taskCanceled) {
 		else {
 			LOG.debug(getLogString("Data source operator cancelled"));
 		}
+
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java]: random new empty line

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java [hunk 8]
```diff
@@ -192,6 +194,13 @@ else if (!this.taskCanceled) {
 			}
 		} finally {
 			BatchTask.clearWriters(eventualOutputs);
+			// --------------------------------------------------------------------
+			// Closing
+			// --------------------------------------------------------------------
+			if(this.format!=null && RichInputFormat.class.isAssignableFrom(this.format.getClass())) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java]: missing space after if and around !=

File: flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSourceBase.java [hunk 9]
```diff
@@ -210,10 +210,12 @@ public void accept(Visitor<Operator<?>> visitor) {
 	protected List<OUT> executeOnCollections(RuntimeContext ctx, ExecutionConfig executionConfig) throws Exception {
 		@SuppressWarnings("unchecked")
 		InputFormat<OUT, InputSplit> inputFormat = (InputFormat<OUT, InputSplit>) this.formatWrapper.getUserCodeObject();
+		//open the input format
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSourceBase.java]: this comment seems out-dated with the differentiation of configure() and openInputFormat()

File: flink-core/src/main/java/org/apache/flink/api/common/io/ReplicatingInputFormat.java [hunk 10]
```diff
@@ -133,4 +133,18 @@ public RuntimeContext getRuntimeContext(){
...
+	@Override
+	public void openInputFormat() {
+		if(this.replicatedIF instanceof RichInputFormat) {
+			((RichInputFormat)this.replicatedIF).openInputFormat();
+		}
+	}
+	
+	@Override
+	public void closeInputFormat() {
+		if(this.replicatedIF instanceof RichInputFormat) {
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/io/ReplicatingInputFormat.java]: missing space after if

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/DataSourceTaskTest.java [hunk 11]
```diff
@@ -255,6 +268,18 @@ public Record readRecord(Record target, byte[] record, int offset, int numBytes)
 			target.setField(1, this.value);
 			return target;
 		}
+		
+		public void openInputFormat() {
+			//ensure this is called only once
+			Assert.assertFalse("Invalid status of the input format. Expected for opened: false, Actual: "+opened, opened);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/DataSourceTaskTest.java]: missing spaces around + (and again in closeInputFormat

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/DataSourceTaskTest.java [hunk 12]
```diff
@@ -84,17 +84,26 @@ public void testDataSourceTask() {
...
 			System.err.println(e);
 			Assert.fail("Invoke method caused exception.");
 		}
 		
-		Assert.assertTrue("Invalid output size. Expected: "+(keyCnt*valCnt)+" Actual: "+this.outList.size(),
+		try {
+			Field formatField = DataSourceTask.class.getDeclaredField("format");
+			formatField.setAccessible(true);
+			MockInputFormat inputFormat = (MockInputFormat) formatField.get(testTask);
+			Assert.assertTrue("Invalid status of the input format. Expected for closed: true, Actual: " + inputFormat.closed, inputFormat.closed);
```
Comment (fhueske) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/DataSourceTaskTest.java]: The check for `inputFormat.opened` is missing, right?

**Comment by fhueske** [comment]
Hi @fpompermaier, thanks for the quick update. I have one last comment. 
The PR is good to merge once that is resolved. Thanks

**Comment by fpompermaier** [comment]
I added the missing assert, now it should be ok!

**Comment by fhueske** [comment]
Thanks @fpompermaier.

Good to merge

**Comment by zentol** [comment]
Merging.
